setup.cfg

```diff
@@ -100,8 +100,8 @@ install_requires =
     funcsigs>=1.0.0, <2.0.0
     graphviz>=0.12
     gunicorn>=19.5.0, <20.0
-    importlib_metadata~=1.7 # We could work with 3.1, but argparse needs <2
-    importlib_resources~=1.4
+    importlib_metadata>=1.5,<4;python_version<="3.7"
```

Comment:
Do we not need to up the min requirements for argcomplete too?

Reply:
Not really. `argcomplete~=1.10` is the same thing as `argcomplete>=1.10,<2`, and so `1.22` is included in that range.

Reply:
Put differently: that'd be a separate task if someone wanted to update this explicitly. But unless you depend on the bugfixes in later releases, I'd leave this be. No need to make the dependency link tighter.

Reply:
Historical we've updated the intermediary rep, otherwise I thought:

Pip would leave argparse at 1.10, which has a dep of importlib_metadata, but then this new dep would install a newer importlib_metadata, leading to incompatible version errors.

Particularly for "older" pips. Am I misremembering?

Reply:
When you use `pip install -U Apache-airflow` to upgrade airflow, a newer `argcomplete` would be installed, automatically.

In my experience it takes installing additional packages with incompatible dependencies for things to break; newer `pip` versions warn when this happens however. Running `pip check` lets you list the problem packages.

Reply:
Also. The problem might be that in our process (in master merges) we automatically do 'pip install .[all] --upgrade --upgrade-strategy eager` to automatically find set of latest constraints that are matching setup.py + setup.cfg limitations, Then we run pip check and then we run all tests and if it all passes. 

As far as I remember without ~1.* for importlib_resources in install_requires caused a problem as pip first tries to install the highest version matching in `install_requires` only - without looking at extras. Once it installs this version, it will only then try to do eager update on all other extras and one of our dependencies already has importib_resources ~= 1.4. Eager upgrade will never downgrade the packages installed by install_requires, so this results in version conflict.

Reply:
Thinking about this some more: we did loosen the restrictions on what versions are acceptable but `importlib_metadata` *is* a new dependency relative to 2.0.0b3. If you then use `pip install apache-airflow>2.0.0b3` (say) *without* `-U` then I think you could indeed end up with an importlib_metadata that is incompatible with an existing argcomplete version if you only have argparse 1.10.

Want me to up the version on that package too, then?

Reply:
In order to test it out, I've added the labels "upgrade to latest dependencies" and "full tests needed". Can you please @mjpieters amend and push --force-with-lease to test my theory? 

With those labels, the PR will automatically try to do what master merge does - it will try to figure out latest constraints with eager upgrade strategy and it will do it for all python versions, so let's see if it works.

Reply:
BTW. We do not recommend running unconstrained upgrade of airflow - this has caused us terrible problems in the past. That's why we have this constraint mechanism and with the recent changes (I am also fixing that for 1.10.14) we take care about making sure that the set of constraints that we recommend (in INSTALL.txt) are non-conflicting and tested (and we update the constraints-1-10 branch and constrains-master automatically to include latest releases of dependencies (all tests muss pass and pip check must show no-conflicts for those as of Friday)

Reply:
Confirmed: @mjpieters @

importlib_resources~=1.4 is a requirement of cnf-lint which is requirement of moto 1.3.16 which we have. So anything above 1 will cause version conflict. mot 1.3.16 is the latest available version, so I do not think we can do anything about it.

```
moto==1.3.16
  - aws-xray-sdk [required: >=0.93,!=0.96, installed: 2.6.0]
    - botocore [required: >=1.11.3, installed: 1.18.18]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
        - six [required: >=1.5, installed: 1.15.0]
      - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
    - future [required: Any, installed: 0.18.2]
    - jsonpickle [required: Any, installed: 1.4.1]
      - importlib-metadata [required: Any, installed: 1.7.0]
        - zipp [required: >=0.5, installed: 3.4.0]
    - wrapt [required: Any, installed: 1.12.1]
  - boto [required: >=2.36.0, installed: 2.49.0]
  - boto3 [required: >=1.9.201, installed: 1.15.18]
    - botocore [required: >=1.18.18,<1.19.0, installed: 1.18.18]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
        - six [required: >=1.5, installed: 1.15.0]
      - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
    - s3transfer [required: >=0.3.0,<0.4.0, installed: 0.3.3]
      - botocore [required: >=1.12.36,<2.0a.0, installed: 1.18.18]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
        - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
          - six [required: >=1.5, installed: 1.15.0]
        - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
  - botocore [required: >=1.12.201, installed: 1.18.18]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
    - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
      - six [required: >=1.5, installed: 1.15.0]
    - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
  - cfn-lint [required: >=0.4.0, installed: 0.42.0]
    - aws-sam-translator [required: >=1.25.0, installed: 1.31.0]
      - boto3 [required: ~=1.5, installed: 1.15.18]
        - botocore [required: >=1.18.18,<1.19.0, installed: 1.18.18]
          - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
          - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
            - six [required: >=1.5, installed: 1.15.0]
          - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
        - s3transfer [required: >=0.3.0,<0.4.0, installed: 0.3.3]
          - botocore [required: >=1.12.36,<2.0a.0, installed: 1.18.18]
            - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
            - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
              - six [required: >=1.5, installed: 1.15.0]
            - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
      - jsonschema [required: ~=3.2, installed: 3.2.0]
        - attrs [required: >=17.4.0, installed: 20.3.0]
        - importlib-metadata [required: Any, installed: 1.7.0]
          - zipp [required: >=0.5, installed: 3.4.0]
        - pyrsistent [required: >=0.14.0, installed: 0.17.3]
        - setuptools [required: Any, installed: 50.3.2]
        - six [required: >=1.11.0, installed: 1.15.0]
      - six [required: ~=1.15, installed: 1.15.0]
    - importlib-resources [required: ~=1.4, installed: 1.5.0]
```

Reply:
Yet again pip's lack of a fully featured dependency solver bites us

Reply:
> In order to test it out, I've added the labels "upgrade to latest dependencies" and "full tests needed". Can you please @mjpieters amend and push --force-with-lease to test my theory?

I rebased and pushed before I noticed that a bot had removed the tags again.

I've pushed a change now that ensures we only allow `importlib-resources` 1.x (starting with `1.4`)

Reply:
yeah - full tests needed only works after the approval status is "Green" (whch is actually good - we can see if it wiorks for 3.6 first and once it does, we approve and the label will be added and we rerun it again with full matrix of tests.

Reply:
Let's see if it works. I will approve it tentatively, and please rebase @mjpieters - let's see if it will work.